stl/inc/ranges

```diff
@@ -1433,6 +1463,94 @@ namespace ranges {
         inline constexpr _Istream_fn<_Ty> istream;
     } // namespace views
 
+#if _HAS_CXX23
+    namespace _Pipe {
+        template <class _Ty>
+        void _Derived_from_range_adaptor_closure_impl(const volatile range_adaptor_closure<_Ty>&);
+
+        template <class _Ty>
+        concept _Derived_from_range_adaptor_closure = requires(const volatile _Ty& __obj) {
+            _Derived_from_range_adaptor_closure_impl(__obj);
+        };
+
+        template <class _Ty>
+        concept _Is_range_adaptor_common =
+            derived_from<remove_cvref_t<_Ty>, range_adaptor_closure<remove_cvref_t<_Ty>>> //
+            && _Derived_from_range_adaptor_closure<_Ty> && !range<_Ty>;
+
+        template <class _Rng, class _Ty>
+        concept _Is_invocable_range_adaptor_closure =
+            range<_Rng> && _Can_pipe<_Rng, _Ty> && _Is_range_adaptor_common<_Ty>;
+
+        template <class _Left, class _Right>
+        concept _Are_composable_range_adaptor_closures =
+            _Is_range_adaptor_common<_Left> && _Is_range_adaptor_common<_Right> && _Can_compose<_Left, _Right>;
+    } // namespace _Pipe
+
+    template <class _Derived>
+        requires is_class_v<_Derived> && same_as<_Derived, remove_cv_t<_Derived>>
+    class range_adaptor_closure {
+    public:
+        template <class _Other>
+            requires _Pipe::_Are_composable_range_adaptor_closures<_Derived, _Other>
+        constexpr auto operator|(_Other&& _Right) && noexcept(
+            noexcept(_Pipe::_Pipeline{static_cast<_Derived&&>(*this), _STD forward<_Other>(_Right)})) {

```

Comment:
Note that this `static_cast` is _not_ just a `std::move`, it also casts to derived class (since `*this` is a `range_adaptor_closure&`)

No change requested.